### ISSUE TITLE
50% performance improvement

### ISSUE THREAD

**Issue by ghost** [serbanghita/Mobile-Detect#424]
Regular expressions consume more time in string long, the idea is to shorten the chain User Agent removed the irrelevant words:

<pre>
$userAgent1 = 'Mozilla/5.0 (Linux; U; Android 4.0.3; ko-kr; LG-L160L Build/IML74K) AppleWebkit/534.30 (KHTML, like Gecko) Version/4.0 Mobile Safari/534.30';
$userAgent2 = '(Android 4.0.3; ko-kr; LG-L160L Build/IML74K) AppleWebkit/534.30  Version/4.0 Mobile Safari/534.30';
$userAgent3 = '(Android 4.0.3; LG-L160L Build/IML74K)  Version/4.0 Mobile Safari/534.30';
$userAgent4 = 'Android 4.0.3; LG-L160L Build/IML74K)  Mobile';

$detect = new Mobile_Detect(null, $userAgent1);
for ( $i = 0; $i < 100; $i++ ) {
  $detect->isMobile();
}
...
userAgent1 267.6149 ms.
userAgent2 184.5591 ms.
userAgent3 139.5360 ms.
userAgent4  90.3389 ms.
</pre>

Developers can do preg_replace in the user agent in the constructor, but  require to know the code and deduce which are the relevant words. In addition to check if it is Tablet irrelevant words may vary from Mobile, or OS version...

In short, simply by doing this:

<pre>
$userAgent = preg_replace('/Mozilla|Linux/', '', $userAgent);
</pre>

Gets a considerable improvement in performance.

**Comment by serbanghita** [comment]
@FranBar interesting point of view, let me expand a bit.

1. For `$detect->isMobile()` and `$detect->isTablet()` methods, yes we could do that. Eliminating **some** of the text pieces that don't help in the detection might help.

2. The action (1) should not affect the original User-Agent `$detect->getUserAgent()`, since we need that for later use, such as detecting browser type and version, os type and version, etc.

Still, the performance optimization should be compared with the extra `preg_replace()` that you mentioned, not on the already stripped User-Agents.

**Comment by FranBar** [comment]
Yes, that is why I have not proposed a patch, it is complex.

It's creating a string with a user agent adapted to each check, which can be created in the constructor and only replaced once, and not in a loop:

$userAgentForMobileCheck =... / / exclude irrelevant words for this case
$userAgentForTabletCheck =... / / exclude irrelevant words for this case
$userAgentForOsCheck =... / / exclude irrelevant words for this case

checkIsMobile ($userAgentForMobileCheck);
checkIsTablet ($userAgentForTabletCheck);

**Comment by FranBar** [comment]
Honestly, I've been using another utility much worse only because the performance of this was poor.

Every day come new devices sometime the performance deteriorate completely.

So I thought maybe I can help with something in this sense, it is a great job that only need to improve performance.

**Comment by FranBar** [comment]
I decided to implement a class:
<pre>
class Mobile_Detect_Fast extends Mobile_Detect
{
    protected $isMobile = null;
    protected $isTablet = null;
    protected $isPhone  = null;
    protected $UAforMobile;
    protected $UAforTablet;
    protected $irrelevant = array(
        'forMobile' => array(
            'Safari.[\d\.]*',
            'Firefox.[\d\.]*',
            'Chrome.[\d\.]*',
            'Chromium.[\d\.]*',
            'MSIE.[\d\.]+',
            'Mozilla.[\d\.]+',
            'AppleWebKit.[\d\.]*',
            'Trident.[\d\.]*',
            'Windows NT.[\d\.]*',
            'Windows',
            'Ubuntu',
            'Linux',
            'Gecko\/\d{8}',
            'KHTML',
            'like Gecko',
            'compatible',
            'x86_..',
            'X11',
            'rv:[\d\.]*',
        ),
        'forTablet' => array(
            'Safari.[\d\.]*',
            'Firefox.[\d\.]*',
            'Chrome.[\d\.]*',
            'Chromium.[\d\.]*',
            'MSIE.[\d\.]+',
            'Mozilla.[\d\.]+',
            'AppleWebKit.[\d\.]*',
            'Trident.[\d\.]*',
            'Windows NT.[\d\.]+',
            'Windows',
            'Ubuntu',
            'Linux',
            'Gecko\/\d{8}',
            'KHTML',
            'like Gecko',
            'compatible',
            'x86_..',
            'X11',
            'rv:[\d\.]*',
        ),
    );

    public function __construct(array $headers = null, $userAgent = null)
    {
        parent::__construct($headers, $userAgent);

        $regexForMobile    = '/'.implode('|', $this->irrelevant['forMobile']).'/';
        $regexForTablet    = '/'.implode('|', $this->irrelevant['forTablet']).'/';
        $this->UAforMobile = preg_replace($regexForMobile, '', $this->userAgent);
        $this->UAforTablet = preg_replace($regexForTablet, '', $this->userAgent);

        // sometimes the user agent is something like "( ; , 2.1 )" in that case it is not necessary check
        if (!preg_match('/\w/', $this->UAforMobile)) {
            $this->isMobile = false;
            $this->isPhone  = false;
        }
        if (!preg_match('/\w/', $this->UAforTablet)) {
            $this->isTablet = false;
        }
    }

    public function isMobileFast($httpHeaders = null)
    {
        if ($this->isMobile === null) {
            $this->isMobile = $this->isMobile($this->UAforMobile, $httpHeaders);
        }

        return $this->isMobile;
    }

    public function isTabletFast($httpHeaders = null)
    {
        if ($this->isTablet === null) {
            $this->isTablet = $this->isTablet($this->UAforTablet, $httpHeaders);
        }

        return $this->isTablet;
    }

    public function isPhoneFast($httpHeaders = null)
    {
        if ($this->isPhone === null) {
            if ($this->isMobileFast($httpHeaders) && !$this->isTabletFast($httpHeaders)) {
                $this->isPhone = true;
            } else {
                $this->isPhone = false;
            }
        }

        return $this->isPhone;
    }

}
</pre>

Test:
<pre>
// get from https://techblog.willshouse.com/2012/01/03/most-common-user-agents/
$userAgents = array(
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/600.6.3 (KHTML, like Gecko) Version/8.0.6 Safari/600.6.3',
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.3; WOW64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko',
    'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (iPhone; CPU iPhone OS 8_3 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12F70 Safari/600.1.4',
    'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (iPad; CPU OS 8_3 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12F69 Safari/600.1.4',
    'Mozilla/5.0 (Windows NT 6.1; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/600.5.17 (KHTML, like Gecko) Version/8.0.5 Safari/600.5.17',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.125 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/600.6.3 (KHTML, like Gecko) Version/7.1.6 Safari/537.85.15',
    'Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; rv:11.0) like Gecko',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/600.5.17 (KHTML, like Gecko) Version/8.0.6 Safari/600.6.3',
    'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko',
    'Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Windows NT 5.1; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/600.7.12 (KHTML, like Gecko) Version/8.0.7 Safari/600.7.12¡',
    'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0)',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/43.0.2357.81 Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/600.4.10 (KHTML, like Gecko) Version/8.0.4 Safari/600.4.10',
    'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/600.7.11 (KHTML, like Gecko) Version/8.0.7 Safari/600.7.11',
    'Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/6.0)',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_8) AppleWebKit/534.59.10 (KHTML, like Gecko) Version/5.1.9 Safari/534.59.10',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:31.0) Gecko/20100101 Firefox/31.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_5) AppleWebKit/537.78.2 (KHTML, like Gecko) Version/6.1.6 Safari/537.78.2',
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.152 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/600.3.18 (KHTML, like Gecko) Version/8.0.3 Safari/600.3.18',
    'Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36',
    'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (Windows NT 6.2; WOW64; rv:38.0) Gecko/20100101 Firefox/38.0',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/41.0.2272.76 Chrome/41.0.2272.76 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.65 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.81 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64; rv:31.0) Gecko/20100101 Firefox/31.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/600.7.10 (KHTML, like Gecko) Version/8.0.7 Safari/600.7.10',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64; rv:31.0) Gecko/20100101 Firefox/31.0 Iceweasel/31.7.0',
    'Mozilla/5.0 (en-US) AppleWebKit/537.36 (KHTML, like Gecko; Hound) Chrome/27.0.1453 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.152 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:37.0) Gecko/20100101 Firefox/37.0',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:39.0) Gecko/20100101 Firefox/39.0',
    'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:37.0) Gecko/20100101 Firefox/37.0',
 );


$timer = microtime(true);
foreach ($userAgents as $ua) {
    $detect = new Mobile_Detect_Fast(null, $ua);
    $resultNative[] = $detect->isMobile();
}
$timeNative = round((microtime(true) - $timer) * 1000, 4);

$timer = microtime(true);
foreach ($userAgents as $ua) {
    $detect = new Mobile_Detect_Fast(null, $ua);
    $resultFast[] = $detect->isMobileFast();
}
$timeFast = round((microtime(true) - $timer) * 1000, 4);
    
echo "native: $timeNative ms. / fast: $timeFast ms. ";
// Check the two methods are the same result
if ( $resultNative == $resultFast ) {
    echo "(same)";
} else {
    echo '(Error: not same)';
}
</pre>

Result:
<pre>
native: 459.6701 ms. / fast: 113.517 ms. (same)
</pre>

**Comment by FranBar** [comment]
To use the parent class method, the native:
<pre>
$detect = new Mobile_Detect_Fast();
$result = $detect->isMobile();
</pre>

To use faster:
</pre>
    $detect = new Mobile_Detect_Fast();
    $result = $detect->isMobileFast();
</pre>

**Comment by FranBar** [comment]
Update:
<pre>
class Mobile_Detect_Fast extends Mobile_Detect
{
    protected $isMobile = null;
    protected $isTablet = null;
    protected $isPhone  = null;
    protected $UAforMobile;
    protected $UAforTablet;
    protected $irrelevant = array(
        'forMobile' => array(
            'Safari.[\d\.]*',
            'Firefox.[\d\.]*',
            'Chrome.[\d\.]*',
            'Chromium.[\d\.]*',
            'MSIE.[\d\.]*',
            'Opera\/[\d\.]*',
            'Mozilla.[\d\.]*',
            'AppleWebKit.[\d\.]*',
            'Trident.[\d\.]*',
            'Windows NT.[\d\.]*',
            'Macintosh.',
            'Windows',
            'Ubuntu',
            'Linux',
            'Intel',
            'Mac OS X',
            'Gecko.[\d\.]*',
            'KHTML',
            'like Gecko',
            'compatible',
            'x86_..',
            'i686',
            'x64',
            'X11',
            'rv:[\d\.]*',
            'Version.[\d\.]*',
            'WOW64',
            'Win64',
            '\.NET CLR [\d\.]*',
            'Presto.[\d\.]*',
        ),
        'forTablet' => array(
            'Safari.[\d\.]*',
            'Firefox.[\d\.]*',
            'Chrome.[\d\.]*',
            'Chromium.[\d\.]*',
            'MSIE.[\d\.]*',
            'Opera\/[\d\.]*',
            'Mozilla.[\d\.]*',
            'AppleWebKit.[\d\.]*',
            'Trident.[\d\.]*',
            'Windows NT.[\d\.]*',
            'Macintosh.',
            'Windows',
            'Ubuntu',
            'Linux',
            'Intel',
            'Mac OS X',
            'Gecko.[\d\.]*',
            'KHTML',
            'like Gecko',
            'compatible',
            'x86_..',
            'i686',
            'x64',
            'X11',
            'rv:[\d\.]*',
            'Version.[\d\.]*',
            'WOW64',
            'Win64',
            '\.NET CLR [\d\.]*',
            'Presto.[\d\.]*',
        ),
    );

    public function __construct(array $headers = null, $userAgent = null)
    {
        parent::__construct($headers, $userAgent);
        $regexForMobile    = '/'.implode('|', $this->irrelevant['forMobile']).'/';
        $regexForTablet    = '/'.implode('|', $this->irrelevant['forTablet']).'/';
        $this->UAforMobile = preg_replace($regexForMobile, '', $this->userAgent);
        $this->UAforTablet = preg_replace($regexForTablet, '', $this->userAgent);
        // sometimes the user agent is something like "( ; , 2.1 )" in that case it is not necessary check
        if (!preg_match('/\w/', $this->UAforMobile) && strpos($this->userAgent, 'Mobile') === false) {
            $this->isMobile = false;
            $this->isPhone  = false;
        }
        if (!preg_match('/\w/', $this->UAforTablet) && strpos($this->userAgent, 'Tablet') === false ) {
            $this->isTablet = false;
        }
    }

    public function isMobileFast($httpHeaders = null)
    {
        if ($this->isMobile === null) {
            if ( strpos($this->userAgent, 'Mobile') !== false ) {
                $this->isMobile = true;
            } else {
                $this->isMobile = $this->isMobile($this->UAforMobile, $httpHeaders);
            }
        }
        return $this->isMobile;
    }

    public function isTabletFast($httpHeaders = null)
    {
        if ($this->isTablet === null) {
            if ( strpos($this->userAgent, 'Tablet') !== false ) {
                $this->isTablet = true;
            } else {
                $this->isTablet = $this->isTablet($this->UAforTablet, $httpHeaders);
            }
        }
        return $this->isTablet;
    }

    public function isPhoneFast($httpHeaders = null)
    {
        if ($this->isPhone === null) {
            if ($this->isMobileFast($httpHeaders) && !$this->isTabletFast($httpHeaders)) {
                $this->isPhone = true;
            } else {
                $this->isPhone = false;
            }
        }
        return $this->isPhone;
    }

}
</pre>
